pkg/planner/cardinality/selectivity.go

```diff
@@ -620,35 +635,86 @@ func GetUsableSetsByGreedy(nodes []*StatsNode) (newBlocks []*StatsNode) {
 				marked[i] = true
 				continue
 			}
-			// We greedy select the stats info based on:
-			// (1): The stats type, always prefer the primary key or index.
-			// (2): The number of expression that it covers, the more the better.
-			// (3): The number of columns that it contains, the less the better.
-			// (4): The selectivity of the covered conditions, the less the better.
-			//      The rationale behind is that lower selectivity tends to reflect more functional dependencies
-			//      between columns. It's hard to decide the priority of this rule against rule 2 and 3, in order
-			//      to avoid massive plan changes between tidb-server versions, I adopt this conservative strategy
-			//      to impose this rule after rule 2 and 3.
-			if (bestTp == ColType && set.Tp != ColType) ||
-				bestCount < bits ||
-				(bestCount == bits && bestNumCols > set.numCols) ||
-				(bestCount == bits && bestNumCols == set.numCols && bestSel > set.Selectivity) {
-				bestID, bestCount, bestTp, bestNumCols, bestMask, bestSel = i, bits, set.Tp, set.numCols, curMask, set.Selectivity
+			current := &statsNodeForGreedyChoice{
+				StatsNode:  set,
+				idx:        i,
+				coverCount: bits,
+			}
+			if current.isBetterThan(best) {
+				best = current
+				bestMask = curMask
 			}
 		}
-		if bestCount == 0 {
+		if best.coverCount == 0 {
 			break
 		}
 
-		// Update the mask, remove the bit that nodes[bestID].mask has.
+		// Update the mask, remove the bit that nodes[best.idx].mask has.
 		mask &^= bestMask
 
-		newBlocks = append(newBlocks, nodes[bestID])
-		marked[bestID] = true
+		newBlocks = append(newBlocks, nodes[best.idx])
+		marked[best.idx] = true
 	}
 	return
 }
 
+type statsNodeForGreedyChoice struct {
+	*StatsNode
+	idx        int
+	coverCount int
+}
+
+func (s *statsNodeForGreedyChoice) isBetterThan(other *statsNodeForGreedyChoice) bool {
+	// none of them should be nil
+	if s == nil || other == nil {
+		return false
+	}
+	// 1. The stats type, always prefer the primary key or index.
+	if s.Tp != ColType && other.Tp == ColType {
+		return true
+	}
```

Comment:
should we insert a 1.5 that `if s.Tp == ColType and other.Tp != ColType { return false }` to skip the remained comparision?

Reply:
The old logic also lacks this check. So I just kept the existing behavior.
And also, after stats v2, this rule is not always correct.